### PR TITLE
Bugfix/v0.12.5

### PR DESCRIPTION
**Pull request by fivetran-reneeli**
## PR Overview
**This PR will address the following Issue/Feature:**
https://github.com/fivetran/dbt_quickbooks/issues/122
https://github.com/fivetran/dbt_quickbooks/issues/80
Zendesk ticket 200711

**This PR will result in the following new package version:**

v0.13.0

**Please provide the finalized CHANGELOG entry which details the relevant changes included in this PR:**

## 🚨 Breaking Changes 🚨:
[PR #124](https://github.com/fivetran/dbt_quickbooks/pull/124) includes the following updates:

## Bug Fixes
- Updates the [int_quickbooks__invoice_join](models/intermediate/int_quickbooks__invoice_join.sql) and downstream [quickbooks__ap_ar_enhanced](dbt_quickbooks/models/quickbooks__ap_ar_enhanced.sql) models to include the `using_payments` config. Previously, this model would fail if the `payment` or the `payment_line` source tables did not exist.
- Requires [quickbooks__ap_ar_enhanced](dbt_quickbooks/models/quickbooks__ap_ar_enhanced.sql) to have `using_payments` enabled in order to run, otherwise the model is limited in its use.
- Updates the [quickbooks__profit_and_loss](models/quickbooks__profit_and_loss.sql) model to include both `period_first_day` and `period_last_day`. This allows users to have greater optionality in choosing which date to aggregate records upon.
- Updates the logic for `amount` in [int_quickbooks__invoice_double_entry](dbt_quickbooks/models/double_entry_transactions/int_quickbooks__invoice_double_entry.sql) to use `invoice.total_amount` only on the condition there exists `bundle_id` and `total_amount = 0`, otherwise use `invoice_lines.amount`. This avoids double counting when aggregating invoice_line items and accounts for the edge cases where a bundle_id is involved.

### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt run –full-refresh && dbt test
- [ ] dbt run (if incremental models are present) && dbt test

Before marking this PR as "ready for review" the following have been applied:
- [x] The appropriate issue has been linked, tagged, and properly assigned
- [x] All necessary documentation and version upgrades have been applied
    
- [x] docs were regenerated (unless this PR does not include any code or yml updates)
- [x] BuildKite integration tests are passing
- [x] Detailed validation steps have been provided below

### Detailed Validation
Please share any and all of your validation steps:

- Customer for #122 has confirmed the `amount` update works
- Issue #80 is straightforward with adding new fields in quickbooks__profit_and_loss
- Zendesk ticket 200711 involved adding `using_payments` config to select models, and this was validated by running with the following vars enabled/disabled:

```
  using_payment: true
  using_bill: true
  using_invoice: true
```

### If you had to summarize this PR in an emoji, which would it be?
 
:dancer:

### CODE REVIEW

File: models/quickbooks__ap_ar_enhanced.sql [hunk 1]
```diff
@@ -143,7 +145,18 @@ final as (
             else 0
                 end as days_overdue,
         invoice_join.initial_payment_date,
-        invoice_join.recent_payment_date
+        invoice_join.recent_payment_date,
+        invoice_join.total_current_payment as total_current_payment
+
+        {% else %}
+        ,
+        null as is_overdue,
```
Comment (fivetran-reneeli) [inline models/quickbooks__ap_ar_enhanced.sql:153]: These changes address the downstream dependency issue when `using_payments` is disabled. Currently I have it so these impacted fields are null if `using_payment` is disabled.

File: models/intermediate/int_quickbooks__invoice_join.sql [hunk 2]
```diff
@@ -21,6 +21,7 @@ estimates as (
 ),
 {% endif %}
 
+{% if var('using_payment', True) %}
```
Comment (fivetran-joemarkiewicz) [inline models/intermediate/int_quickbooks__invoice_join.sql:24]: The `int_quickbooks__invoice_join` model is only used as an upstream model in the `quickbooks__ap_ar_enhanced` model.

image

Additionally, this model is not materialized and is [ephemeral in our config](https://github.com/fivetran/dbt_quickbooks/blob/9f759070707abe0971ee7199335abca61b391a7c/dbt_project.yml#L17-L18). 

Because of this, would it make more sense to take the same approach you took for the `quickbooks__ap_ar_enhanced` model and just entirely disable this model if payments are disabled? This would allow us to reduce the code changes and not run code in the customers warehouse if it is not entirely necessary.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,13 @@
...
+- Requires [quickbooks__ap_ar_enhanced](dbt_quickbooks/models/quickbooks__ap_ar_enhanced.sql) to have `using_payments` enabled in order to run, otherwise the model is limited in its use.
+- Updates the [quickbooks__profit_and_loss](models/quickbooks__profit_and_loss.sql) model to include both `period_first_day` and `period_last_day`. This allows users to have greater optionality in choosing which date to aggregate records upon.
+- Updates the logic for `amount` in [int_quickbooks__invoice_double_entry](dbt_quickbooks/models/double_entry_transactions/int_quickbooks__invoice_double_entry.sql) to use `invoice.total_amount` only on the condition there exists `bundle_id` and `total_amount = 0`, otherwise use `invoice_lines.amount`. This avoids double counting when aggregating invoice_line items and accounts for the edge cases where a bundle_id is involved.
```
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:9]: Not all of these are breaking changes. Would you be able to separate out which are breaking and which are general bugfixes.

Comment (fivetran-joemarkiewicz) [reply CHANGELOG.md:9]: Additionally, be sure to include any other updates that came from this review.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,13 @@
+# dbt_quickbooks v0.13.0
+## 🚨 Breaking Changes 🚨:
+[PR #124](https://github.com/fivetran/dbt_quickbooks/pull/124) includes the following updates:
+
+## Bug Fixes
+- Updates the [int_quickbooks__invoice_join](models/intermediate/int_quickbooks__invoice_join.sql) and downstream [quickbooks__ap_ar_enhanced](dbt_quickbooks/models/quickbooks__ap_ar_enhanced.sql) models to include the `using_payments` config. Previously, this model would fail if the `payment` or the `payment_line` source tables did not exist.
+- Requires [quickbooks__ap_ar_enhanced](dbt_quickbooks/models/quickbooks__ap_ar_enhanced.sql) to have `using_payments` enabled in order to run, otherwise the model is limited in its use.
```
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:7]: These two bullets are essentially saying the same thing. Is there a way to consolidate or streamline?

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,13 @@
...
+- Updates the [quickbooks__profit_and_loss](models/quickbooks__profit_and_loss.sql) model to include both `period_first_day` and `period_last_day`. This allows users to have greater optionality in choosing which date to aggregate records upon.
+- Updates the logic for `amount` in [int_quickbooks__invoice_double_entry](dbt_quickbooks/models/double_entry_transactions/int_quickbooks__invoice_double_entry.sql) to use `invoice.total_amount` only on the condition there exists `bundle_id` and `total_amount = 0`, otherwise use `invoice_lines.amount`. This avoids double counting when aggregating invoice_line items and accounts for the edge cases where a bundle_id is involved.
+
```
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:10]: Small wording suggestions here
```suggestion
- Updates the logic for the `amount` field in [int_quickbooks__invoice_double_entry](dbt_quickbooks/models/double_entry_transactions/int_quickbooks__invoice_double_entry.sql) to use `invoice.total_amount` only on the condition when a bundle is associated with the invoice and `invoice.total_amount` is 0, otherwise `invoice_lines.amount` is used. 
   - This avoids double counting when aggregating invoice_line items and accounts for the edge cases where a bundle_id is involved.

```

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-reneeli great job working through these changes and confirming the updates work for the customer. I do have a few comments and suggestions before approving. Let me know if you have any questions. Thanks!

File: dbt_project.yml [hunk 6]
```diff
@@ -1,7 +1,7 @@
 config-version: 2
 name: 'quickbooks'
 
-version: '0.12.4'
+version: '0.13.0'
```
[the remark this answers is not in the document]
Comment (fivetran-reneeli) [reply dbt_project.yml:4]: thanks for catching! updated

File: models/quickbooks__profit_and_loss.sql [hunk 7]
```diff
@@ -8,6 +8,8 @@ with general_ledger_by_period as (
 final as (
     select
         period_first_day as calendar_date,
+        period_first_day,
+        period_last_day,
```
[the remark this answers is not in the document]
Comment (fivetran-reneeli) [reply models/quickbooks__profit_and_loss.sql:12]: Ah good point, added that and updated the changelog to include

File: models/quickbooks__ap_ar_enhanced.sql [hunk 8]
```diff
@@ -126,13 +126,12 @@ final as (
         concat(billing_address.address_1, billing_address.address_2) as customer_vendor_address_line,
         {% endif %}
 
-        customers.website as customer_vendor_webiste,
+        customers.website as customer_vendor_website,
```
[the remark this answers is not in the document]
Comment (fivetran-reneeli) [reply models/quickbooks__ap_ar_enhanced.sql:129]: updated and separated out the breaking changes

File: models/intermediate/int_quickbooks__invoice_join.sql (hunk 2 above)
Comment (fivetran-reneeli) [reply models/intermediate/int_quickbooks__invoice_join.sql:24]: Ah I think this is a good point-- There's no point in running `int_quickbooks__invoice_join` if the downstream `quickbooks__ap_ar_enhanced` is disabled. I added the same configs (invoice and payments enabled) to this model.

File: CHANGELOG.md (hunk 3 above)
Comment (fivetran-reneeli) [reply CHANGELOG.md]: Updated!

File: CHANGELOG.md (hunk 4 above)
Comment (fivetran-reneeli) [reply CHANGELOG.md:7]: consolidated!

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,15 @@
+# dbt_quickbooks v0.13.0
+[PR #124](https://github.com/fivetran/dbt_quickbooks/pull/124) includes the following updates:
+
+## 🚨 Breaking Changes 🚨:
+- Updates the [int_quickbooks__invoice_join](https://github.com/fivetran/dbt_quickbooks/blob/main/models/intermediate/int_quickbooks__invoice_join.sql) and downstream [quickbooks__ap_ar_enhanced](https://github.com/fivetran/dbt_quickbooks/blob/main/models/quickbooks__ap_ar_enhanced.sql) models to include and require the `using_payments` config. Previously, these models would fail if the `payment` or the `payment_line` source tables did not exist.
+- Corrects the misspelled `customer_vendor_webiste` to `customer_vendor_website` in `quickbooks__ap_ar_enhanced`.
+
+## Bug Fixes
+- Updates the [quickbooks__profit_and_loss](https://github.com/fivetran/dbt_quickbooks/blob/main/models/quickbooks__profit_and_loss.sql)and [quickbooks__balance_sheet](https://github.com/fivetran/dbt_quickbooks/blob/main/models/quickbooks__balance_sheet.sql) models to include both `period_first_day` and `period_last_day`. This allows users to have greater optionality in choosing which date to aggregate records upon.
```
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:9]: Small wording suggestions
```suggestion
- Updates the [quickbooks__profit_and_loss](https://github.com/fivetran/dbt_quickbooks/blob/main/models/quickbooks__profit_and_loss.sql) and [quickbooks__balance_sheet](https://github.com/fivetran/dbt_quickbooks/blob/main/models/quickbooks__balance_sheet.sql) models to include both `period_first_day` and `period_last_day`. This allows users to have greater flexibility in choosing which date to aggregate records upon.
```

File: models/intermediate/int_quickbooks__invoice_join.sql (hunk 2 above)
Comment (fivetran-joemarkiewicz) [reply models/intermediate/int_quickbooks__invoice_join.sql:24]: Looks great! Technically, you can revert your changes from before where the specific `{% if var('using_payment', True) %}` conditionals are included. Now that this model won't be run there is no need for these added conditionals.

Comment (fivetran-joemarkiewicz) [reply models/intermediate/int_quickbooks__invoice_join.sql:24]: You can safely use the previous code from this model, but only make the following change and that should return the desired results
```sql
{{ config(enabled=var('using_invoice') and var('using_payment', True)) }}
```

**Review by fivetran-joemarkiewicz — COMMENTED** [review]
@fivetran-reneeli great work on this PR! I have two final suggestions, but they don't need to hold back the approval. Once you make those final changes and regen the docs, you can open this for release review!

Let me know if you have any questions. Thanks!

**Review by fivetran-joemarkiewicz — APPROVED** [review]
See the above comment for approval notes

File: models/intermediate/int_quickbooks__invoice_join.sql (hunk 2 above)
Comment (fivetran-reneeli) [reply models/intermediate/int_quickbooks__invoice_join.sql:24]: ah good point-- updated to revert and remove those original changes

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,15 @@
+# dbt_quickbooks v0.13.0
+[PR #124](https://github.com/fivetran/dbt_quickbooks/pull/124) includes the following updates:
+
+## 🚨 Breaking Changes 🚨:
+- Updates the [int_quickbooks__invoice_join](https://github.com/fivetran/dbt_quickbooks/blob/main/models/intermediate/int_quickbooks__invoice_join.sql) and downstream [quickbooks__ap_ar_enhanced](https://github.com/fivetran/dbt_quickbooks/blob/main/models/quickbooks__ap_ar_enhanced.sql) models to include and require the `using_payments` config. Previously, these models would fail if the `payment` or the `payment_line` source tables did not exist.
+- Corrects the misspelled `customer_vendor_webiste` to `customer_vendor_website` in `quickbooks__ap_ar_enhanced`.
```
Comment (fivetran-jamie) [inline CHANGELOG.md:6]:
```suggestion
- Corrects the misspelled `customer_vendor_webiste` field to `customer_vendor_website` in `quickbooks__ap_ar_enhanced`.
```

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,3 +1,15 @@
...
+
+## Bug Fixes
+- Updates the [quickbooks__profit_and_loss](https://github.com/fivetran/dbt_quickbooks/blob/main/models/quickbooks__profit_and_loss.sql) and [quickbooks__balance_sheet](https://github.com/fivetran/dbt_quickbooks/blob/main/models/quickbooks__balance_sheet.sql) models to include both `period_first_day` and `period_last_day`. This allows users to have greater flexibility in choosing which date to aggregate records upon.
```
Comment (fivetran-jamie) [inline CHANGELOG.md:9]: is this a bug fix or feature update?

Comment (fivetran-jamie) [reply CHANGELOG.md:9]: also are we intending on deprecating `calendar_date`? if so, maybe we should mention it here (and maybe in `calendar_date`'s yml description) and recommend using one of the new fields instead

File: models/quickbooks__balance_sheet.sql [hunk 12]
```diff
@@ -8,6 +8,8 @@ with general_ledger_by_period as (
 final as (
     select
         period_first_day as calendar_date,
+        period_first_day,
```
Comment (fivetran-jamie) [inline models/quickbooks__balance_sheet.sql:11]: maybe we should note here as well in a comment that calendar_date will be deprecated

File: models/quickbooks__profit_and_loss.sql [hunk 13]
```diff
@@ -8,6 +8,8 @@ with general_ledger_by_period as (
 final as (
     select
         period_first_day as calendar_date,
+        period_first_day,
```
Comment (fivetran-jamie) [inline models/quickbooks__profit_and_loss.sql:11]: same comment as above

**Review by fivetran-jamie — COMMENTED** [review]
looks great! just some doc-related comments around deprecating `calendar_date`

File: CHANGELOG.md (hunk 11 above)
Comment (fivetran-reneeli) [reply CHANGELOG.md:9]: Thanks @fivetran-jamie , moved that to its own feature update section and added the deprecation notice for `calendar_date`:

"This is slated to be deprecated, and the fields `period_first_day` and `period_last_day` are both offered as replacements, depending on how your company performs their financial reporting."

File: models/quickbooks__balance_sheet.sql (hunk 12 above)
Comment (fivetran-reneeli) [reply models/quickbooks__balance_sheet.sql:11]: good point! added

File: models/quickbooks__profit_and_loss.sql (hunk 13 above)
Comment (fivetran-reneeli) [reply models/quickbooks__profit_and_loss.sql:11]: added

File: CHANGELOG.md (hunk 11 above)
Comment (fivetran-jamie) [reply CHANGELOG.md:9]: that sounds great! do you have the feature_update section added locally?

**Review by fivetran-jamie — APPROVED** [review]

File: models/quickbooks__profit_and_loss.sql [hunk 14]
```diff
@@ -7,7 +7,7 @@ with general_ledger_by_period as (
 
 final as (
     select
-        period_first_day as calendar_date,
+        period_first_day as calendar_date, --  Slated to be deprecated; we recommend using `period_first_day` or `period_last_day`
```
Comment (sbrudz) [inline models/quickbooks__profit_and_loss.sql:10]: I just noticed that period_first_day and period_last_day aren't included in the quickbooks__profit_and_loss SQL even though the yml definition says the columns exist.

Comment (fivetran-reneeli) [reply models/quickbooks__profit_and_loss.sql:10]: Ah @sbrudz you're right, this was mistakenly removed during the PR. We will have this fix ready first thing next week!

Comment (sbrudz) [reply models/quickbooks__profit_and_loss.sql:10]: Thank you!

Comment (fivetran-reneeli) [reply models/quickbooks__profit_and_loss.sql:10]: Thanks @sbrudz  for catching this, the package has been updated!
